File: solutions/0338_Counting_Bits/math.java
class Solution {
    public int[] countBits(int num) {
        int[] ret = new int[num + 1];
        if (num == 0) {
            ret[0] = 0;
            return ret;
        }
        // let f(n) to be the number of 1's in n's binary representation
        // f(n) = f(n / 2) + 1    if n % 2 == 1
        //      = f(n / 2)        otherwise
        for (int i = 1; i < ret.length; ++i) {
            ret[i] = ret[i / 2] + (i % 2);
        }
        return ret;
    }
}

File: solutions/0338_Counting_Bits/math.py
class Solution:
    def countBits(self, num: int) -> List[int]:
        if num == 0:
            return [0]
        # let k to be 2's exponential growth
        # 00000 = 0 2**0
        # 00001 = 1 < 2**1 - 1
        # 00010 = 1
        # 00011 = 2 << 2**2 - 1
        # 00100 = 1
        # 00101 = 2
        # 00110 = 2
        # 00111 = 3 <<< 2**3 - 1
        # 01000 = 1
        # 01001 = 2
        # 01010 = 2
        # 01011 = 3
        # 01100 = 2
        # 01101 = 3
        # 01110 = 3
        # 01111 = 4 <<<< 2**4 - 1
        # 10000 = 1
        if num == 0:
            return [0]
        ret = [0]
        k = 1
        while len(ret) < num + 1:
            ret += [v + 1 for v in ret]
            k += 1
        return ret[:num + 1]
